### DISCUSSION TITLE
OpenGL ES 3.1 support using MoltenVK on macOS

### DISCUSSION THREAD

**Discussion by guanzhangrtk** [KhronosGroup/MoltenVK#1339] [Ideas]
Android Emulators on PC and Mac typically use ANGLE or their own implementation to provide OpenGL ES capability to support apps that require it.

Currently on Windows, graphics rendering comes in two flavours, one via Direct3D (using ANGLE library, limited to OpenGL ES 3.0 support) and OpenGL (using proprietary (?) implementation of OpenGL ES translator via OpenGL 4.5+, this works with games that require OpenGL ES 3.1). On the Mac, rendering is mostly done via ANGLE using presumably the stock Apple OpenGL since the full OpenGL implementation is not available, support for OpenGL ES 3.1 is currently not possible.

I was reading up on the ANGLE project webpage and apparently they have OpenGL ES 3.1 implementation using Vulkan backend, so I was wondering whether it would be possible to add OpenGL ES 3.1 support to macOS using MoltenVK.

If it is indeed possible, would anybody be interested in working on such a project together? As far as I know none of the existing Android Emulators on the Mac has support for OpenGL ES 3.1 due to macOS' OpenGL limitations. Thank you for reading!

**Comment by galad87** [comment]
There is https://github.com/kakashidinho/metalangle running Angle on Metal. I don't know if it implements enough OpenGL ES 3.0 features for you needs.

**Reply by guanzhangrtk** [reply]
Thanks for the link but I think that code base is being migrated to the official ANGLE repo anyway, besides I need 3.1 support not 3.0. This chart from the official ANGLE project shows that Vulkan backend support for 3.1 is done but not for Metal:

![image](https://user-images.githubusercontent.com/44019025/116344088-3a8e0700-a818-11eb-8d2d-b92b7c2a6eed.png)

Thanks for the link anyway!

**Comment by cdavis5e** [comment]
Almost all of what you need should already work. Here's a list of new features in GLES 3.1, relative to 3.0:

* Arrays of arrays - should just work (I think)
* Compute shaders - supported
* Indirect draw commands - supported
* Explicit uniform location - supported
* Framebuffers with no attachments - supported on macOS 10.15, iOS 13.0; emulated by MoltenVK on lower OSes
* Atomic counters - supported (I think)
* Shader bitfield operations - supported
* Shader helper invocation - supported on macOS 10.14 or iOS 14
* Shader image load/store operations - supported, yes even atomics on `R32I` and `R32UI` images
* Shader layout binding - supported
* Shader storage buffer objects - supported
* Separate shader objects - this is just how Vulkan and Metal work
* Stencil texturing - supported
* Texture gather operations - supported
* Multisample formats for immutable textures - supported
* Vertex attribute binding - supported

**Reply by guanzhangrtk** [reply]
@cdavis5e Thanks for your response but can you please elaborate what you mean by "should already work"?

I'm not a programmer, so I was not planning to write OpenGL ES 3.1 code to run on macOS. However I do want to run apps on Android Emulators which require OpenGL ES 3.1 support. At most what I could do is build ANGLE w/ Vulkan backend support and install that along with MoltenVK and hope that it works magically. But perhaps I'm oversimplifying things.

Thanks in advance for any feedback!

**Reply by cdavis5e** [reply]
Oh, sorry. All of that is simply to say that yes, it should just magically work. Well, mostly, anyway, assuming you can get ANGLE to use Vulkan on macOS. If you encounter problems at runtime, feel free to file bugs.

**Reply by guanzhangrtk** [reply]
Oh this is encouraging... I'll try to build ANGLE on macOS and see how far I can go. Really appreciate it!

**Reply by Artein** [reply]
@guanzhangrtk, have you found a way to support OpenGL ES 3.1 on macOS inside Android Emulator?

**Reply by italomandara** [reply]
https://moltengl.com... I hoped this could also have an update to support newer versions of opengl-es at some point, it would be much better than going through angle -> vulkan -> metal supposedly.

**Comment by yshuolu** [comment]
This is very interesting. Do we have a success try here?

**Comment by mpalomas** [comment]
I am late to the party but I was curious, and the result: it works!

> INFO: GL: OpenGL device information:
> INFO:     > Vendor:   Google Inc. (Apple)
> INFO:     > Renderer: ANGLE (Apple, Vulkan 1.2.296 (Apple M1 Pro (0x0F030207)), MoltenVK-0.2.2019)
> INFO:     > Version:  OpenGL ES 3.1.0 (ANGLE 2.1.25153 git hash: 1b34d2a18af1)
> INFO:     > GLSL:     OpenGL ES GLSL ES 3.10 (ANGLE 2.1.25153 git hash: 1b34d2a18af1)
> INFO: GL: Supported extensions count: 179

That's crazy. I've just run basic Raylib examples to get that, nothing fancy, so I'm not sure it's 100% compliant, but the report says so.

**Reply by GimmiGoose** [reply]
So you got 3.1 working?

**Reply by atilaalcantara** [reply]
did you?
